presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1732,6 +1769,29 @@ else if (criteria instanceof JoinOn) {
             return output;
         }
 
+        private void verifyOnConditionReferencesRelatedFields(Relation rightRelation, Expression expression)
+        {
+            if (!onConditionReferencesRelatedFields(expression, rightRelation)
+                    && (rightRelation instanceof AliasedRelation || rightRelation instanceof Table)) {
```

Comment:
Since we are checking for these types below, we don't need to check for this here, right? This can be simplified alot.

Reply:
This check is needed, because the logic inside the `if` is specifically for the  case `rightRelation` is either `AliasedRelation` or  `Table`. For instance, It will not work for the case when `rightRelation` is `TableSubquery`, `Unnest`,  or `Lateral`.

Reply:
Can you leave a TODO?

Reply:
Added TODO